vpr/src/pack/cluster.cpp

```diff
@@ -1358,6 +1416,10 @@ static enum e_block_pack_status try_pack_molecule(t_cluster_placement_stats* clu
                         revert_place_atom_block(molecule->atom_block_ids[i], router_data, atom_molecules);
                     }
                 }
+
+                /* Placement failed, clean the pb */
```

Comment:
To clarify, is the cleanup_pb function supposed to be called when a molecule fails try_pack_molecule for any reason? Or only when it fails during atom placement (try_place_atom_block_rec)? I’m asking because in the issue it says “Add (or fix?) cleanup code that removes child_pbs created during atom placement if it fails for any reason.”

If it is meant to be called when a molecule fails for any reason, it should also be called if the atom fails the detailed routing legality check in try_pack_molecule. (line 1305/1363)

If the cleanup function is only meant to be called if the atom fails during try_place_atom_block_rec , then maybe the call for cleanup_pb should be at the point right after try_place_atom_block_rec is called. So right after the atom placement attempt, you could have a check of the block_pack_status and call the cleanup function if it needed. Right now, where the cleanup_pb function is called, it could have failed the placement check or the pin feasibility filter check.

Reply:
You're right. It should be called whenever `try_place_atom_block_rec` fails. But if you look at the loop that invokes it for each atom you'll see that it terminates immediately on failure: https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/8e00e82a03ce12b7987a244620a0e3624083ae79/vpr/src/pack/cluster.cpp#L1241

Later in the flow there are calls to some atom placement reversal routines (https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/8e00e82a03ce12b7987a244620a0e3624083ae79/vpr/src/pack/cluster.cpp#L1350) executed on failure so I figured out that the pb cleanup should be invoked from there. It doesn't matter what is the reason of the atom placement failure. If no pbs were allocated the function simply won't do anything. BTW: it does not touch pbs with already placed atoms.

Diving into the low-level details: The problem I observed with the assertion failure in `try_place_atom_block_rec` was that for some reason the function was hitting already allocated pbs with their modes set causing a conflict with the mode required by the currently packed atom. Digging deeper I've discovered that the function is called for a pb_graph_node corresponding to that atom. Then it recurses up the hierarchy allocating pbs and setting their modes until the root CLB is reached. But these pbs are never freed between subsequent molecule packing. So if packing succeeds everything is fine but when it fails there are leftover pbs set to specific modes. This may cause mode mismatch if the mode required for the next molecule is different than the last one checked.

I typed "Add (or fix?)..." in the issue as I was surprised that the problem haven't manifested itself until now. I'm still not sure whether I'm adding the actually missing data structure cleanup or if the cleanup is already there and requires fixing and I've missed it.

Reply:
Ah I see, that makes sense. In that case LGTM.